internal/proxyservice/proxy_service.go

```diff
@@ -1,43 +1,140 @@
 	v2 "github.com/cyberark/secretless-broker/pkg/secretless/config/v2"
-	"github.com/cyberark/secretless-broker/pkg/secretless/log"
+	logapi "github.com/cyberark/secretless-broker/pkg/secretless/log"
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector"
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector/tcp"
 )
 
 // TODO: move to impl package
 type proxyServices struct {
-	config        v2.Config
-	logger        log.Logger
-	eventNotifier v1.EventNotifier
-	availPlugins  plugin.AvailablePlugins
+	config          v2.Config
+	logger          logapi.Logger
+	eventNotifier   v1.EventNotifier
+	availPlugins    plugin.AvailablePlugins
+	runningServices []secretless.Service
 }
 
-// TODO: Rename to Call or Run and return a Stopper instead of having Stop()
 // Start starts all proxy services
-func (s *proxyServices) Start() {
-	// TODO: Implement
-
-	// For each ProxyService:
-	//
-	// 1. Rewrap the Logger with service name prefix
-	// 2. Create the ConnectorResources object
-	// etc...
+func (s *proxyServices) Start() error {
+	for _, svc := range s.servicesToStart() {
+		err := svc.Start()
+		// Showstopper: Failure to start a service is fatal.
+		if err != nil {
+			// TODO: Upgrade our logger so we can use Fatalf here
+			s.logger.Panicf("could not start proxy service: %s", err)
+		}
+		s.runningServices = append(s.runningServices, svc)
+	}
+	return nil
 }
 
 // Stop stops all proxy services
-func (s *proxyServices) Stop() {
-	// Stop all the Service
+func (s *proxyServices) Stop() error {
+	var stopFailures []string
+	for _, svc := range s.runningServices {
+		err := svc.Stop()
+		if err != nil {
+			s.logger.Errorf("could not stop proxy service: %s", err)
+			stopFailures = append(stopFailures, err.Error())
+		}
+	}
+
+	if len(stopFailures) > 0 {
+		return fmt.Errorf(
+			"these errors occured while stopping all services: %s",
+			strings.Join(stopFailures, "; "),
+		)
+	}
+	return nil
+}
+
+func (s *proxyServices) servicesToStart() []secretless.Service {
+	var servicesToStart []secretless.Service
+
+	tcpPlugins := s.availPlugins.TCPPlugins()
+	// httpPlugins := s.availPlugins.HTTPPlugins()
+
+	for _, svc := range s.config.Services {
+		requestedPlugin := svc.Connector //TODO: this rename is a name smell
+
+		// first check the available TCP Plugins
+		tcpPlugin, found := tcpPlugins[requestedPlugin]
+		if found {
+			tcpSvc, err := s.createTCPService(svc, tcpPlugin)
+			// Failure to create is fatal.
+			if err != nil {
+				// TODO: Add Fatalf to our logger and use that
+				s.logger.Panicf("unable to create TCP service '%s'", svc.Name)
+			}
+			servicesToStart = append(servicesToStart, tcpSvc)
+		}
+
+		// TODO: next check available HTTP Plugins
+		// httpPlugin, found := httpPlugins[requestedPlugin]
+
+		// TODO: Deal with SSH in a hardcoded way
+
+		// Default case: not found
+		// TODO: Add Fatalf to our logger and use that
+		s.logger.Panicf("plugin '%s' not available.", requestedPlugin)
```

Comment:
in that case there should be a `continue` for the found case

Reply:
The panic is effectively fatal (see the TODO above) because it is never `recover`ed.  It actually had a `continue` before and I removed it in response to Srdjan's "this is unreachable" comment lol.

Reply:
The issue is that the block at line 73 is executed because the tcp plugin is found but on line 80 we don't `continue` instead we go back to the parent block and print `"plugin '%s' not available."` even though the plugin is available.

Reply:
Ah, sorry I read your initial comment too fast.  Thanks!